### PR TITLE
local rack on kubernetes

### PR DESCRIPTION
**Pull request by ddollar**
Kubernetes-powered local Rack

To test:

```
$ convox update 20190122141721-k8s-local
$ sudo convox rack install local -v 20190122141721-k8s-local
```

### CODE REVIEW

**Comment by qnm** [comment]
Early days yet but I'm unable to run this with microk8s on Ubuntu 18.10.

It installs correctly using `sudo convox rack install local -n test -v 20181110181749-k8s-local` however hangs when `Waiting on rack...`

I note that both the `router` and `resolver` services are stuck in pending.

I see the following event log `pod: "router-5d98cbf96d-rnm5k_convox-system(bc7494fd-e660-11e8-981e-3413e8430cd3)". kubelet does not have ClusterDNS IP configured and cannot create Pod using "ClusterFirst" policy. Falling back to "Default" policy. `

**Comment by ddollar** [comment]
Thanks for trying this on Linux, it's on my todo list for this week. You could help me out a lot with that process (I'm not a native Linux-workstation user, would be doing the VM thing) would be to see if you can figure out what tweaks are necessary to get past that.

One other thing I need to figure out is whether or not NetworkManager allows for alternate ports in its custom DNS setup. I've just guessed at this config file.

https://github.com/convox/rack/pull/3015/files#diff-cb84a270c9148bcb585a6b25d120e809R33

**Comment by qnm** [comment]
> Thanks for trying this on Linux, it's on my todo list for this week. You could help me out a lot with that process (I'm not a native Linux-workstation user, would be doing the VM thing) by seeing if you can figure out what tweaks are necessary to get past that.
>

It looks like it's an artifact of the service type being `LoadBalancer` and no external IP being available.

The endpoints are available and appear active:

```
$ kubectl get endpoints --namespace=convox-system
NAME       ENDPOINTS                    AGE
resolver   10.1.1.32:5453               128m
router     10.1.1.32:80,10.1.1.32:443   128m

$ curl -I 10.1.1.32:80
HTTP/1.1 301 Moved Permanently
Content-Type: text/html; charset=utf-8
Location: https://10.1.1.32/
Date: Tue, 13 Nov 2018 01:22:10 GMT
```
 
> One other thing I need to figure out is whether or not NetworkManager allows for alternate ports in its custom DNS setup. I've just guessed at this config file.
> 
> [rack/provider/local/system_linux.go](https://github.com/convox/rack/blob/756afc0ebc87603ff34ea59610e0ce24fdb8b1ad/provider/local/system_linux.go#L33)
> 
> Line 33 in [756afc0](/convox/rack/commit/756afc0ebc87603ff34ea59610e0ce24fdb8b1ad)
>  data = []byte(fmt.Sprintf("server=/%s/127.0.0.1:5453\n", name))

This looks correct, however I'm not sure how `127.0.0.1` can be used. Either the service has to bind to `127.0.0.1` or use the cluster ip instead. I'm not particularly clued up on k8s so I can't really help.

That said, I'm happy to try things and report back.

**Comment by qnm** [comment]
Installed as 'test'. There's both a 'convox' in there and 'test'.

Let me remove the 'convox' one.

**Comment by qnm** [comment]
I'm seeing `Failed to add DNS server address '127.0.0.1:5453', ignoring: Invalid argument`

Worth noting that it doesn't seem to be possible to have `systemd-resolved` listening on an alternate port, see https://github.com/systemd/systemd/issues/4377 which refers to a similar case with `dnscrypt`.

The suggestion is to bind to port 53 on a different ip in the 127.0.0.0/8 range.

**Comment by ddollar** [comment]
@qnm Try again with the version `20181114021011-k8s-local`

**Comment by qnm** [comment]
Okay, have installed the client and attempted to create the rack.

Doesn't look like systemd-resolved likes the config, though it's reading it:

```
$ journalctl -f -u systemd-resolved
-- Logs begin at Fri 2018-10-26 22:05:10 AEDT. --
Nov 14 13:42:48 glenlivet systemd[1]: Starting Network Name Resolution...
Nov 14 13:42:49 glenlivet systemd-resolved[10777]: Positive Trust Anchors:
Nov 14 13:42:49 glenlivet systemd-resolved[10777]: . IN DS 19036 8 2 49aac11d7b6f6446702e54a1607371607a1a41855200fd2ce1cdde32f24e8fb5
Nov 14 13:42:49 glenlivet systemd-resolved[10777]: . IN DS 20326 8 2 e06d44b80b8f1d39a95c0b0d7c65d08458e880409bbc683457104237c7f8ec8d
Nov 14 13:42:49 glenlivet systemd-resolved[10777]: Negative trust anchors: 10.in-addr.arpa 16.172.in-addr.arpa 17.172.in-addr.arpa 18.172.in-addr.arpa 19.172.in-addr.arpa 20.172.in-addr.arpa 21.172.in-addr.arpa 22.172.in-addr.arpa 23.172.in-addr.arpa 24.172.in-addr.arpa 25.172.in-addr.arpa 26.172.in-addr.arpa 27.172.in-addr.arpa 28.172.in-addr.arpa 29.172.in-addr.arpa 30.172.in-addr.arpa 31.172.in-addr.arpa 168.192.in-addr.arpa d.f.ip6.arpa corp home internal intranet lan local private test
Nov 14 13:42:49 glenlivet systemd-resolved[10777]: Failed to add DNS server address '127.0.0.1:5453', ignoring: Invalid argument
Nov 14 13:42:49 glenlivet systemd-resolved[10777]: Using system hostname 'glenlivet'.
Nov 14 13:42:49 glenlivet systemd[1]: Started Network Name Resolution.
```

Confirming that the resolver is in place and running:

```
$ kubectl get endpoints --namespace=convox-system
NAME       ENDPOINTS                    AGE
resolver   10.1.1.81:5453               27h
router     10.1.1.81:80,10.1.1.81:443   27h

$ nslookup -port=5453 rack.test 10.1.1.81
Server:		10.1.1.81
Address:	10.1.1.81#5453

Name:	rack.test
Address: 127.0.0.1
Name:	rack.test
Address: 127.0.0.1
```

However not available on 127.0.0.1

```
$ nslookup -port=5453 rack.test 127.0.0.1
;; connection timed out; no servers could be reached
```

And not available via systemd-resolved

```
$ nslookup -port=53 rack.test 127.0.0.53
Server:		127.0.0.53
Address:	127.0.0.53#53

** server can't find rack.test: NXDOMAIN
````

**Comment by ddollar** [comment]
It's not on 127.0.0.1. Try `resolvectl query rack.test`

**Comment by ddollar** [comment]
Also make sure you blow away any manual files you created in resolved.d

**Comment by qnm** [comment]
Sorry, still not working.

I've confirmed I've removed my test files from `/etc/systemd/resolved.d` and have since rebooted.

The resolved config is in place

```
$ cat /usr/lib/systemd/resolved.conf.d/convox.test.conf 
[Resolve]
DNS=127.0.0.1:5453
Domains=~test
```

But it won't resolve

```
$ resolvectl query rack.test
rack.test: resolve call failed: 'rack.test' not found
```

**Comment by ddollar** [comment]
Are you sure you have updated your cli and reinstalled the local rack? It should not be showing 127.0.0.1 in that config file

**Comment by qnm** [comment]
Binary looks up-to-date

```
~$ sudo convox -v
[sudo] password for qnm: 
20181114021011-k8s-local
```

File: provider/local/system_linux.go [hunk 1]
```diff
@@ -60,7 +62,7 @@ func dnsInstallNetworkManager(name string) error {
 func dnsInstallResolved(name string) error {
 	data := []byte(fmt.Sprintf("[Resolve]\nDNS=127.0.0.1:5453\nDomains=~%s", name))
```
Comment (qnm) [inline provider/local/system_linux.go]: Should this line use the Cluster IP rather than localhost?

Comment (ddollar) [reply provider/local/system_linux.go]: Yes, you're right. Not sure how I missed that.

**Comment by ddollar** [comment]
New release is `20181114123334-k8s-local`

**Comment by qnm** [comment]
If I remove `1.1.1.1` from my main `/etc/systemd/resolved.conf` file, I can resolve `rack.test` but unfortunately can't resolve anything else.

**Comment by ddollar** [comment]
It works in my VM. On my VM my /etc/systemd/resolved.conf has no configuration in it, only a blank [Resolve] section but internet-host-resolving still works.

```
Global
       LLMNR setting: no
MulticastDNS setting: no
  DNSOverTLS setting: no
      DNSSEC setting: no
    DNSSEC supported: no
  Current DNS Server: 10.152.183.14
         DNS Servers: 10.152.183.14
          DNS Domain: ~abcd
                      ~dev
          DNSSEC NTA: 10.in-addr.arpa
                      16.172.in-addr.arpa
                      168.192.in-addr.arpa
                      17.172.in-addr.arpa
                      18.172.in-addr.arpa
                      19.172.in-addr.arpa
                      20.172.in-addr.arpa
                      21.172.in-addr.arpa
                      22.172.in-addr.arpa
                      23.172.in-addr.arpa
                      24.172.in-addr.arpa
                      25.172.in-addr.arpa
                      26.172.in-addr.arpa
                      27.172.in-addr.arpa
                      28.172.in-addr.arpa
                      29.172.in-addr.arpa
                      30.172.in-addr.arpa
                      31.172.in-addr.arpa
                      corp
                      d.f.ip6.arpa
                      home
                      internal
                      intranet
                      lan
                      local
                      private
                      test

Link 110 (veth26f063a4)
      Current Scopes: none
       LLMNR setting: yes
MulticastDNS setting: no
  DNSOverTLS setting: no
      DNSSEC setting: no
    DNSSEC supported: no

Link 108 (veth7a6a241c)
      Current Scopes: none
       LLMNR setting: yes
MulticastDNS setting: no
  DNSOverTLS setting: no
      DNSSEC setting: no
    DNSSEC supported: no

Link 103 (vethb1b26cff)
      Current Scopes: none
       LLMNR setting: yes
MulticastDNS setting: no
  DNSOverTLS setting: no
      DNSSEC setting: no
    DNSSEC supported: no

Link 101 (vethb226b052)
      Current Scopes: none
       LLMNR setting: yes
MulticastDNS setting: no
  DNSOverTLS setting: no
      DNSSEC setting: no
    DNSSEC supported: no

Link 8 (cbr0)
      Current Scopes: none
       LLMNR setting: yes
MulticastDNS setting: no
  DNSOverTLS setting: no
      DNSSEC setting: no
    DNSSEC supported: no

Link 3 (docker0)
      Current Scopes: none
       LLMNR setting: yes
MulticastDNS setting: no
  DNSOverTLS setting: no
      DNSSEC setting: no
    DNSSEC supported: no

Link 2 (ens33)
      Current Scopes: DNS
       LLMNR setting: yes
MulticastDNS setting: no
  DNSOverTLS setting: no
      DNSSEC setting: no
    DNSSEC supported: no
  Current DNS Server: 192.168.225.2
         DNS Servers: 192.168.225.2
          DNS Domain: ~.
                      localdomain
```

**Comment by qnm** [comment]
I'd say that's because you have your DNS applied at the link

```
Link 2 (ens33)
      Current Scopes: DNS
       LLMNR setting: yes
MulticastDNS setting: no
  DNSOverTLS setting: no
      DNSSEC setting: no
    DNSSEC supported: no
  Current DNS Server: 192.168.225.2
         DNS Servers: 192.168.225.2
          DNS Domain: ~.
                      localdomain
```

I will apply the same change to my setup and test.

**Comment by ddollar** [comment]
That configuration was just from a fresh install of 18.10

**Comment by qnm** [comment]
I installed the minimal version of 18.10, I regret it now as so many pieces are missing. This might be one of them.

I'm going to refresh my install on the weekend, but regardless this appears to be working for me now. Thanks for all your help.

**Comment by slaskis** [comment]
Thanks for the yaml fix!

I've been trying to get my app running with this PR and I wanted to post some findings:

some are restricting me from using it:

- resources does not seem to be accessible. they start and log properly but it seems an ingress is missing or something? even within the app it seems unable to find the resources. no workaround found yet for this. 
- no code sync? or it tries but fails with `sync add error: command terminated with exit code 1`. 

some less important ones:

- Service links does not work. I think this might actually be a non-feature, but it was very handy as the env vars remote and local got setup correctly without me having to look through CF outputs and thus it makes the transition slightly more cumbersome. worked around by setting env vars manually.
- healthchecks work now. this is obviously a good thing but previously it let me ignore that dynamodb-local does not respond to GET requests. worked around with a http proxy.
- docker instability. i don't know why but docker kept crashing after a deploy. but seems to work better after switching to the stable channel.
- default scale results in pods failing. i'm not sure what the limit is in kubernetes but it only managed to start 1 pod before failing with "insufficient memory" errors. worked around by setting scale for each service to cpu=64/mem=64.
- it seems to have missed to uninstall the old router. so now the router logs just output the convox usage message on repeat

**Comment by jayjanssen** [comment]
Testing out k8s local rack and I'm unable to `convox exec` into the ID  of a process that I've `convox start`ed.  I get the error message:

> ERROR: unable to upgrade connection: container not found ("main")

**Pull request closed without merging**